rp2040-hal/src/uart.rs

```diff
@@ -1,3 +1,356 @@
+        Ok(bytes_written)
+    }
+
+    /// Reads bytes from the UART.
+    /// This function reads as long as it can. As soon that the FIFO is empty, if :
+    /// - 0 bytes were read, a WouldBlock Error is returned
+    /// - some bytes were read, it is deemed to be a success
+    /// Upon success, the number of read bytes is returned.
+    pub fn read(&self, buffer: &mut [u8]) -> nb::Result<usize, Infallible> {
+
+        let mut bytes_read = 0;
+
+        Ok(loop {
+            if !self.uart_is_readable() {
+                if bytes_read == 0 {
+                    return Err(WouldBlock)
+                }
+                else {
+                    return Ok(bytes_read)
+                }
+            }
+
+            if bytes_read < buffer.len() {
+                buffer[bytes_read] = self.device.uartdr.read().data().bits();
+                bytes_read += 1;
+            }
+            else {
+                break bytes_read;
+            }
+        })
+    }
+
+    /// Writes bytes to the UART.
+    /// This function blocks until the full buffer has been sent.
+    pub fn write_full_blocking(&self, data: &[u8]) {
+
+        let mut offset = 0;
+
+        while offset != data.len() {
+            offset += match self.write(&data[offset..]) {
+                Ok(written_bytes) => {
+                    written_bytes
+                }
+
+                Err(WouldBlock) => continue,
+
+                Err(_) => unreachable!()
+            };
+        }
+    }
+
+    /// Reads bytes from the UART.
+    /// This function blocks until the full buffer has been received.
+    pub fn read_full_blocking(&self, buffer: &mut [u8]) {
+        let mut offset = 0;
+
+        while offset != buffer.len() {
+            offset += match self.read(&mut buffer[offset..]) {
+                Ok(bytes_read) => {
+                    bytes_read
+                }
+
+                Err(WouldBlock) => continue,
+
+                Err(_) => unreachable!()
+            };
+        }
+    }
+
+}
+
+/// Baudrate configuration. Code loosely inspired from the C SDK.
+fn configure_baudrate(device: &mut dyn UARTDevice, wanted_baudrate: &Baud, frequency: &Hertz) -> Baud {
+
+    let frequency = *frequency.integer();
+
+    let baudrate_div = 8 * frequency / *wanted_baudrate.integer();
+
+    let (baud_ibrd, baud_fbrd) = match (baudrate_div >> 7, ((baudrate_div & 0x7F) + 1) / 2) {
+
+        (0, _) => (1, 0),
+
+        (ibrd, _) if ibrd >= 65535 => (65535, 0),
+
+        (ibrd, fbrd) => (ibrd, fbrd)
+    };
+
+    // Load PL011's baud divisor registers
+    device.uartibrd.write(|w| unsafe {
+        w.baud_divint().bits(baud_ibrd as u16);
+        w
+    });
+    device.uartfbrd.write(|w| unsafe {
+        w.baud_divfrac().bits(baud_fbrd as u8);
+        w
+    });
+
+    // PL011 needs a (dummy) line control register write to latch in the
+    // divisors. We don't want to actually change LCR contents here.
+    device.uartlcr_h.write(|w| {
+        w
+    });
```

Comment:
I think write will overrite all the values. Maybe you meant `modify()`?

Reply:
I asked myself the same question when writing this. I'm not sure what happens when `write()` is called without actually making any changes to the register. I also tested with `modify()` and with nothing at all (commented out the code) and it does not seem to have any impact. This code is there only because I tried to mimick what's being done in the C SDK.
What they actually do is :
```c
    // PL011 needs a (dummy) line control register write to latch in the
    // divisors. We don't want to actually change LCR contents here.
    hw_set_bits(&uart_get_hw(uart)->lcr_h, 0);
```
So they write 0 ? I'm not sure what this does and how it can be translated using our PAC.

Reply:
As I understand it `.write()` only gives you the reset value of the register so you have a struct to call the methods for each field on. I'm not sure what `hw_set_bits` does but I would assume that it just is a `register |= 0` which writes to the register but with the original value. In Rust that should be `.modify(|_, w| w)`.